odoo/addons/rpc2/__init__.py

```diff
@@ -0,0 +1,253 @@
+# -*- coding: utf-8 -*-
+import base64
+import collections
+import json
+import logging
+import threading
+import types
+from datetime import datetime
+try:
+    from xmlrpc import client as xmlrpclib
+except ImportError:
+    import xmlrpclib
+
+import werkzeug.exceptions
+import werkzeug.wrappers
+from lxml import etree
+from lxml.builder import E
+
+from odoo import http, models, service, registry
+from odoo.service import security
+
+from . import global_, database, model
+from . import models
+
+logger = logging.getLogger(__name__)
+
+
+class Rpc2(http.Controller):
+    @http.route('/RPC2/', auth='none', methods=['POST'], csrf=False)
+    def rpc2(self, db=None):
```

Comment:
Using py3 `xmlrpc.client` only the path `/RPC2/` is used to send the http request, the query `?db=...` is discared. I opened a bug report in cpython https://bugs.python.org/issue43330

Reply:
WRT fragment, that's normal, the fragment is never sent to servers.

WRT query string however, that's annoying. And even if a fix is merged it won't be backported to every version under the sun, so might be necessary to change the scheme. Apparently at one point it was `/RPC2/<dbname>`, sadly I've no idea why I changed it to `/RPC2?db=<dbname>`. Looking at the examples, none of the libraries seems to have special support for that form over the other. I can only guess it was "better semantics".

Reply:
We would have to wait a bit anyway, there is the standard `functools.singledispatchmethod` that can be used to simplify the custom marshaller

Reply:
> We would have to wait a bit anyway, there is the standard functools.singledispatchmethod that can be used to simplify the custom marshaller

That's a completely different concern: `singledispatch` is a server-side simplification (so it's not actually necessary in any way). We control the server, we can do that whenever we want.

The issue here is *client-side*, we don't control that and saying that clients have to use 3.9 or 3.10 would be stupid.

Reply:
Not that waiting for `singledispatchmethod` is necessary either, the serialisation system is not really overridable so it could be built out of functions, passing the `memo` set explicitly instead of implicitly.